lib/utility/sink.rb

```diff
@@ -12,28 +12,57 @@
 require 'utility/es_client'
 
 module Utility
-  module Sink
-    module_function
+  class Batcher
+    def initialize(sink, flush_threshold = 50)
+      # threshold and size are the same now, but it's good to separate the concepts:
+      # threshold is "when to start flushing", while "size" is the size of the batch
+      # that will be sent to the sink
+      @flush_threshold = flush_threshold
+      @flush_size = flush_threshold
+      @sink = sink
+      @queue = []
+    end
+
+    def add(element)
+      @queue << element
+
+      flush if ready_to_flush?
+    end
+
+    def add_multiple(elements)
+      elements.each do |element|
+        @queue << element
+      end
 
-    def print_delim
-      puts '----------------------------------------------------'
+      flush if ready_to_flush?
```

Comment:
should move this into `each` block?

Reply:
Yeah probably a good idea, need to think more about it.
Just thinking out loud:

The difference will be in behaviour when `ingest_multiple` is used, for example connector attempts to ingest 60 documents, while batching is set up to send data to Elasticsearch in batches of 50.

In currently done way it will do the following: add 60 elements to the queue, then send 50 + 10 elements to Elasticsearch immediately after.

If we do the change proposed by @wangch079, it will send 50 elements to Elasticsearch, then add 10 more and wait.

Intuitively, second case looks better. Any additional thoughts?

Reply:
I have a revolutionary idea (kidding).

```
    def add(element)
      @queue << element

      flush if ready_to_flush?
    end

    def add_multiple(elements)
      elements.each do |element|
        add(element)
      end
    end
```

Why do you add the single-item method if not to use it? =)